PersistentVolumeClaimBinder

markturansky opened the pull request:
Following #6002 (client).  
From #5105.

markturansky commented:
@derekwaynecarr you can ignore all these client commits on this PR.   They come by "soft resetting" the client branch to my manager branch.  They will drop off once the client is merged.

This manager implementation can use some cleanup after the client is merged and before I ask for a review.

markturansky commented:
@thockin This is ready for review.

The two big additions are:

* PersistentVolumeManager is a simple "sync the world" loop like every other controller.  I would like to refactor this into @lavalamp's DeltaFIFO queue as a follow-on task.
* custom cache.Store/Index in pkg/volumemanager/types.go that indexes PVs by access modes and keeps them sorted for easy matching.

markturansky commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -190,6 +194,9 @@ func (s *CMServer) Run(_ []string) error {
 	namespaceManager.Run(s.NamespaceSyncPeriod)
 
+	persistentVolumeManager := volumemanager.NewPersistentVolumeManager(kubeClient)
+	persistentVolumeManager.Run(s.NamespaceSyncPeriod)
```
wrong period variable.

thockin commented on pkg/api/v1beta1/types.go:
```diff
@@ -122,6 +122,8 @@ type PersistentVolumeSource struct {
 	// kubelet's host machine and then exposed to the pod.
 	GCEPersistentDisk *GCEPersistentDiskVolumeSource `json:"persistentDisk" description:"GCE disk resource provisioned by an admin"`
+	// NFS represents an NFS mount on the host that persists beyond a pod's lifetime
+	NFS *NFSVolumeSource `json:"nfs" description:"NFS volume that will be mounted in the host machine "`
```
nit: all the NFS deltas can be a different PR

markturansky replied:
Do you mind if I leave them in?  I added them so I could test the indexing of different access modes on volumes.  My tests create GCE and NFS example volumes for indexing, sorting, and searching.

I can change it to HostPath, if you prefer, to similarly test indexing by access modes.  I thought NFS made for a better test.

thockin replied:
But we want NFS for real, right?  Why not just send that as a PR now, and then you can depend on it (and this one gets a smidge smaller).

markturansky replied:
We do want NFS as a real persistent volume.  I can send it along as a new PR for a quick merge, if you prefer that.  Then I can use that in my tests or I can change my tests to use HostPath.  I don't need NFS specifically for access modes that differ from GCE.  

I have no preference.

thockin replied:
If we want this anyway and it is totally unrelated to this actual PR, send it and I'll merge it quickly.

markturansky replied:
happily, Gluster was recently merged and I picked it up after a rebase.  I'll use that one instead.

thockin commented on pkg/volumemanager/types.go:
```diff
@@ -0,0 +1,173 @@
+	indexer cache.Indexer
+}
+
+var _ = cache.Store(&persistentVolumeOrderedIndex{}) // persistentVolumeOrderedIndex is a Store
```
Minor: The pattern I grok better is 

```
var _ cache.Store = &persistentVolumeOrderedIndex{}
```

thockin commented on pkg/volumemanager/types.go:
```diff
@@ -0,0 +1,173 @@
+
+var _ = cache.Store(&persistentVolumeOrderedIndex{}) // persistentVolumeOrderedIndex is a Store
+
+func (pvstore *persistentVolumeOrderedIndex) Add(obj interface{}) error {
```
Why wrap all of these, rather than just embed the cache.Indexer and get this for free?

markturansky replied:
Ah, that must have been what Derek was referring to in another comment.

thockin commented on pkg/volumemanager/types.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+// AccessModesIndexFunc is an indexing function that returns a persistent volume's AccessModes as a string
+func AccessModesIndexFunc(obj interface{}) (string, error) {
```
doesn't need to be exported

thockin commented on pkg/volumemanager/types.go:
```diff
@@ -0,0 +1,173 @@
+	return "", fmt.Errorf("object is not a persistent volume: %v", obj)
+}
+
+// List returns all volumes with the given set of AccessModeTypes *in order* of their storage capacity (low to high)
```
Godoc comment should start with the symbol name

thockin commented on pkg/volumemanager/types.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+// MatchPredicate is a function that indicates that a persistent volume matches another
+type MatchPredicate func(compareThis, toThis *api.PersistentVolume) bool
```
doesn't need to be exported

thockin commented on pkg/volumemanager/types.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+// onlyUnboundedVolumes is a MatchPredicate that filters bound volumes before comparing storage capacity
+func onlyUnboundedVolumes(compareThis, toThis *api.PersistentVolume) bool {
```
S/Unbounded/Unbound

Unbounded means "without limits", Unbound means "without binding"

thockin commented on pkg/volumemanager/types.go:
```diff
@@ -0,0 +1,173 @@
+	if toThis.Spec.ClaimRef != nil || compareThis.Spec.ClaimRef != nil {
+		return false
+	}
+	return matchStorageCapacity(compareThis, toThis)
```
pattern-wise, this function is not self-describing, either rename to something like matchUnboundCapacity() or something.  Since it is only used in one place, you might just inline it.

markturansky replied:
I also use that func in the comparator that orders the volumes.   This func doesn't know anything about bound or unbound volumes.

thockin replied:
Sorry, I meant this function "onlyUnboundedVolumes" is not self describing.

thockin commented on pkg/volumemanager/types.go:
```diff
@@ -0,0 +1,173 @@
+		volumes[i] = obj.(*api.PersistentVolume)
+	}
+
+	comparator := persistentVolumeComparator{volumes}
```
minor: a pattern I have seen all over and sort of like for readability is to name the comparator type "byCapacity", so you get

```
sort.Sort(byCapacity(volumes))
```

thockin commented on pkg/volumemanager/persistent_volume_manager.go:
```diff
@@ -0,0 +1,154 @@
+		claimStore:  claimStore,
+		client:      kubeClient,
+	}
+	manager.syncHandler = manager.syncPersistentVolumeClaim
```
Why is this outside of the literal block above it?

markturansky replied:
Cargo cult.  All the other controllers are the same.

markturansky replied:
Probably because we need an instance of the "PVMgr" first before we can access it for the sync func?   can't self-reference inside the literal block.

thockin replied:
ah, right, I missed that this was a self-referential method pointer (even though it doesn't curry the 'this', sigh).

thockin commented on pkg/volumemanager/persistent_volume_manager.go:
```diff
@@ -0,0 +1,154 @@
+			return fmt.Errorf("Unexpected error getting volume reference: %v\n", err)
+		}
+
+		volume.Spec.ClaimRef = claimRef
```
Can you remind me why ClaimRef is Spec and not Status?

markturansky replied:
Because @bgrant0607 said so :)

Anything persisted needs to be on Spec whereas Status is recreated by observation, even though my observers are persisting status to the new /status subresource.

thockin replied:
I disagree with that distinction, but I it's not the time or place :)  OK

thockin commented on pkg/volumemanager/persistent_volume_manager.go:
```diff
@@ -0,0 +1,154 @@
+limitations under the License.
+*/
+
+package volumemanager
```
some docs on this pkg?

markturansky replied:
Sure.  While I'm at it, name change?

I originally thought I'd need to sync 2 objects.  I don't, just claims for binding.  The custom store that indexes and sorts volumes is sufficient and the reflector keeps that cache populated.

Is this control loop just a ```PersistentVolumeClaimBinder``` ?

thockin replied:
Name change if you think it makes sense.  Where do you intend the dynamic provisioning to live?

markturansky replied:
In a separate controller that's watching levels of volumes, as opposed to making this do 2 things.

thockin commented on pkg/volumemanager/persistent_volume_manager.go:
```diff
@@ -0,0 +1,154 @@
+		return nil
+	}
+
+	volume, err := controller.volumeStore.FindByClaim(claim)
```
FindByClaim makes it sound like you're looking for an extant match.  But this is really an allocation, right?  Maybe AllocateByClaim() or FindBestByClaim() or something?

thockin commented on pkg/volumemanager/persistent_volume_manager.go:
```diff
@@ -0,0 +1,154 @@
+		volume.Spec.ClaimRef = claimRef
+		claim.Status.VolumeRef = volumeRef
+
+		volume, err = controller.client.PersistentVolumes().Update(volume)
```
I am not deeply familiar with the client libs yet - will this do resourceVersion checking and fail if someone updated the object underneath you?  Maybe a comment to that effect.

markturansky replied:
I have validation in my client code to check for a resourceVersion and I confirmed with @derekwaynecarr that the server does, indeed, check for this.

thockin commented on pkg/volumemanager/persistent_volume_manager.go:
```diff
@@ -0,0 +1,154 @@
+		} else {
+			_, err = controller.client.PersistentVolumeClaims(claim.Namespace).Update(claim)
+			if err != nil {
+				glog.V(3).Infof("Error updating claim: %+v\n", err)
```
If there's an error here, won't you leave the volume with a live ClaimRef but the Claim without a VolumeRef?

markturansky replied:
I'd give your left arm for multi-object transactions in etcd.

thockin replied:
Should you un-update the PV?

markturansky replied:
Yes, it's un-updated now.

markturansky commented:
@thockin All feedback addressed.  Also rebased.

Need to implement DeltaFIFO for the controller to pick up delete events for claims.  There is no other way for a volume to know it was unbound (except a "sync the world" loop for PVs).

markturansky commented:
For expediency, I reverted to the original working "sync the world" loops similar to the other controllers and I'd like to refactor this into the newer controller framework by @lavalamp in a subsequent task.

There is a bug/data race in Go 1.3 that I'm investigating.

The unit tests work with the /examples and they bind correctly.

thockin commented:
re-reviewing, needs rebase

thockin commented on pkg/api/types.go:
```diff
@@ -204,6 +204,8 @@ type VolumeSource struct {
 
 // Similar to VolumeSource but meant for the administrator who creates PVs.
 // Exactly one of its members must be set.
+// All PersistentVolumes are volumes, only some Volumes are persistent.  All pointers
```
I'm not sure the last part of this comment is true for very long.   I would just drop this whole comment - it doesn't add much for me.

markturansky replied:
I agree.  I've been thinking about how the pointer must be in both VS and PVS.

I think I have to create the entire PersistentVolumePlugin implementation as completely separate from VolumePlugin.  This would allow the pointer to be in just PSV, leaving VS for other purposes.

thockin replied:
You need to distinguish the implementation from the API.

In the API, PVSource and VSource will be partially overlapping sets.  In the implementation PLugin is a strict subset of VPlugin (or else you could not SetUp).  For example, we need AWSEBS or GCEPD to be a VolumePlugin, because we want to use it as a PersistentVolume, but we don't want to offer it directly as a VolumeSource (in the future).

markturansky replied:
In that case, do we need an internal type to resolve the standoff between VS and PVS?  Refactor the plugins to work against the internal type.  The API types translate to it.  Is this a weird thing to express in the converters and internal types.go file?

thockin replied:
as per the other PR, yes, this is a valid option, but it doesn't need to be captured in the API at all.

markturansky replied:
I think we're saying the same thing.  VS and PVS are in the API and contain whatever pointers to volumes as appropriate.  There is an internal type (not in the API) that contains all the pointers.  VS and PVS translate to the internal type.  Volume plugins work with the internal type.  Is that correct?

markturansky replied:
Maybe disregard the last comment here?  I'm crossing the streams and just read your comments on the other PR.  Must acquire mutex comment lock.

thockin replied:
I think we've converged

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,216 @@
+limitations under the License.
+*/
+
+package volumeclaimbinder
```
regarding pkgs - does it make sense to put all this into pkg/persistentvolume or something?  It's minor, but this feels like a component of some larger things..

markturansky replied:
This binder (and the proposed PersistentVolumeControllerManager for provisioning) are intended to be controller plugins.  I think packaging them beneath /controllers/{plugin-name} like Volumes makes sense.

thockin replied:
What is the plan for controller plugins?

markturansky replied:
I can package this anywhere you'd like in the meantime, until we have plugins.

I assume, though don't know, that controller plugins would be packaged like volume plugins.

markturansky commented:
@thockin I'd prefer to redo this binder in the newer plugin framework designed by @lavalamp and made into plugins (https://github.com/GoogleCloudPlatform/kubernetes/pull/6520).

If you'd review the volume plugin (https://github.com/GoogleCloudPlatform/kubernetes/pull/6695) and the Controllers Plugin PR (above), I will redo this controller in the other framework for review.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,216 @@
+			//claim was deleted by user.
+			glog.V(3).Infof("PersistentVolumeClaim[UID=%s] unbound from PersistentVolume[UID=%s]\n", volume.Spec.ClaimRef.UID, volume.UID)
+			volume.Spec.ClaimRef = nil
+			volume.Status.Phase = api.VolumeReleased
```
Something about this irks me.  Maybe not a big deal, but I need to write it.  Status is supposed to be "rediscoverable" in general.  Yet, if we somehow lost Status here, we could not differentiate a PV that was used and released from one that was ready to use.  Is it important to remember who used to claim this volume?  e.g. instead of blanking ClaimRef, set a flag or something?  Or maybe just leave it in this state (appears to be claimed, claim does not exist) until the recycler comes along?

markturansky replied:
In the case of lost status, we can infer the PV is released by looking up (and not finding) the PVC claim reference.

With /status as a persistent subresource on PV, I would not be adverse to storing the authoritative binding ClaimRef there.  This the the API discussion you need to have with @bgrant0607.

thockin replied:
My point was that if Status is supposed to always be re-observable (we are not consistent in this) then in this case Spec will have ClaimRef = nil, but it is in a different state than a new PV whose ClaimRef = nil.

My suggestion being, why not set Phase to Released, but leave the ClaimRef.  That way, even if Status somehow got lost (imagine we wanted to store it less durably since it is supposed to be re-observable), this loop would be able to re-determine that it was Released.

markturansky replied:
Makes sense.  Leaving the ref allows recreation of the "released" status when the volume verifies its claim check no longer exists.  Future provisioning tasks can rely on this status being set at some point.

thockin replied:
After discussing with @bgrant0607 and others by email (kubernetes-dev) I think I was correct.  You should leave the claim populated and set the Phase and the recycler can use the Phase to trigger recycles, at which point Claimref is cleared.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,216 @@
+		if err != nil {
+			glog.V(3).Infof("Error updating volume: %+v\n", err)
+		} else {
+
```
useless blank line

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,216 @@
+
+				// uset ClaimRef on the pointer to make it available for binding again
+				volume.Spec.ClaimRef = nil
+				volume, err = controller.client.PersistentVolumes().Update(volume)
```
re-check err?

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder_test.go:
```diff
@@ -0,0 +1,200 @@
+
+func TestExampleObjects(t *testing.T) {
+
+	scenarios := map[string]struct {
```
minor: why the struct here, rather than map[string]interface{} ?  I'm actually a bit surprised the syntax works - any reason for this?

markturansky replied:
It's an inline definition that I learned from other unit tests.  I've been making all my test scenarios like this across all unit tests.  It's pretty convenient.

thockin replied:
Just curious - how is it more convenient?

```
scenarios := map[string]struct {
    expected interface{}
}{
    "foo": {
        expected: &Bar{...}
    }
}
```

vs

```
scenarios := map[string]interface{}{
    "foo": &Bar{...}
}
```

markturansky replied:
What is &Bar{} in this example?  Do I have to define Bar's properties elsewhere in the test file?  

The structs I define inline like this contain all the properties related to that specific testing scenario.

thockin replied:
concretely, taking your code

```
scenarios := map[string]struct {
		expected interface{}
}{
	"claims/claim-01.yaml": &api.PersistentVolumeClaim{
		Spec: api.PersistentVolumeClaimSpec{
			AccessModes: []api.AccessModeType{api.ReadWriteOnce},
			Resources: api.ResourceRequirements{
				Requests: api.ResourceList{
					api.ResourceName(api.ResourceStorage): resource.MustParse("3Gi"),
				},
			},
		},
	},
```

The only difference is that I omitted "{\n\t\t\texpected: " and " one closing curly.

markturansky replied:
Oh, I see now.   In this particular test I had but the one property in each "scenario" whereas all my other tests usually have many properties in the struct.  I just kept the same pattern.  I see how I can reduce it here if I have no other properties to add to the test scenarios.

thockin replied:
You don't have to fix it, I just wanted to understand if there was some trickery going on that I did not understand.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder_test.go:
```diff
@@ -0,0 +1,200 @@
+		t.Error("Expected to find volume in the index")
+	}
+
+	err = controller.syncPersistentVolumeClaim(claim)
```
What about testing the other half (unbinding when claim is removed)

thockin commented on pkg/volumeclaimbinder/types.go:
```diff
@@ -0,0 +1,147 @@
+type matchPredicate func(compareThis, toThis *api.PersistentVolume) bool
+
+// Find returns the nearest PV from the ordered list or nil if a match is not found
+func (pvstore *persistentVolumeOrderedIndex) Find(pv *api.PersistentVolume, matchPredicate matchPredicate) (*api.PersistentVolume, error) {
```
does this need to be exported symbol?

thockin replied:
s/pvstore/index/ or something (all over)

markturansky commented:
I have no idea how this just got closed ... I force pushed my cleaned up branch.  Will open another.

markturansky committed:
PersistentVolumeClaimBinder implementation

markturansky commented:
nm

markturansky committed:
Addressed feedback, improved flow and comments

markturansky committed:
goformatted

markturansky committed:
renamed exported symbol for pvOrderedIndex, added locks, remove go routine for sync funcs

markturansky commented:
@thockin feedback addressed except for the 2nd half of the test (delete a claim).  I made #6928 to help with that piece.

There is a data race in the tests for Go 1.3 that I am still investigating.

markturansky committed:
narrowed client interface to allow easier testing.  added PVC deletion test case.

markturansky commented:
@thockin ready for review.

Refactored to a narrow client interface (similar to @ironcladlou's recent work #6871) which made testing much easier.  Full binding/release unit tests implemented.  The examples work on the CLI and I'll edit the README tomorrow.

markturansky committed:
Edited README

markturansky commented:
Something with the reflectors in this setup is still wonky in Go 1.3.   This binder works fine on my macbook (Go 1.4) but fails to update a PV's Status in Go 1.3 on F20.

I am refactoring to the new framework.Informer framework by @lavalamp to hopefully avoid all my threading issues.

markturansky committed:
Refactored to DeltaFifo

markturansky committed:
Added more to README, tweaked pod, added service, fixed validation

markturansky commented:
@thockin ready for review.  

* Refactored into a DeltaFifo implementation which avoids my earlier threading bugs.
* Added smarter "should I update" logic
* Fixed a small couple of issues
* The README is an epic piece of prose

markturansky committed:
Added integration test, fixed a validation issue

markturansky commented:
Added integration tests that uses a running binder for matching.

ironcladlou commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+				syncClaimStatus(binderClient, volume, claim)
+			}
+		} else {
+			//claim was deleted by user.
```
I don't think it's safe to assume that without checking the error type (e.g. `errors.IsNotFound`).

thockin replied:
+1

ironcladlou commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+		volume.Spec.ClaimRef = claimRef
+		volume, err = binderClient.UpdatePersistentVolume(volume)
+
+		if err != nil {
```
As above, you probably want a stronger error assertion here.

markturansky replied:
@ironcladlou Do we care here, in this case?  Any error here is unrecoverable, nils the reference just made, and returns.

ironcladlou replied:
Resource conflicts are recoverable.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+			volume.Status.Phase = api.VolumeBound
+			_, err := binderClient.UpdatePersistentVolumeStatus(volume)
+			if err != nil {
+				return fmt.Errorf("Error updating pv.status: %v\n", err)
```
Is it correct to leave volume.Status.Phase set to Bound ?

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+			volume.Status.Phase = api.VolumeReleased
+			volume, err = binderClient.UpdatePersistentVolumeStatus(volume)
+			if err != nil {
+				return fmt.Errorf("Error updating pv: %+v\n", err)
```
Is it correct to leave Phase as Released?

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+	claim.Status.Capacity = volume.Spec.Capacity
+
+	_, err = binderClient.UpdatePersistentVolumeClaimStatus(claim)
+
```
blank line

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+		return err
+	}
+
+	if volume != nil {
```
flip this to be if volume == nil first, handle error and return.  undent the rest.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+
+		if err != nil {
+			// volume no longer bound
+			volume.Spec.ClaimRef = nil
```
Don't you need to call binderClient.UpdatePersistentVolume(volume) again?

markturansky replied:
If the update failed with an error, this resets the value of the pointer back to what it was previously.  It should match what is in etcd.  Is this an incorrect assumption?

thockin replied:
Sorry, brain error.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+			// volume no longer bound
+			volume.Spec.ClaimRef = nil
+			return fmt.Errorf("Error updating volume: %+v\n", err)
+		} else {
```
don't need the else block

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+		} else {
+			err = syncClaimStatus(binderClient, volume, claim)
+			if err != nil {
+				return fmt.Errorf("Error update claim.status: %+v\n", err)
```
nit: s/update/updating

need to rollback prior txn here?

markturansky replied:
The rollback happens in the same syncClaimStatus func, because I use that func twice and didn't want to duplicate the rollback.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,301 @@
+
+		// make a binding reference to the claim
+		volume.Spec.ClaimRef = claimRef
+		volume, err = binderClient.UpdatePersistentVolume(volume)
```
Covering this whole block - I am not sure in reading this that rollback of this multi-object transaction is right.  I know it's an annoying corner case, but I think it needs to be clearer what is happening and that this is on purpose.

Really good comments would help.

Being more explicit about what gets reverted would help.

Something like a defer'ed rollback or something would make it clearer.  Maybe something like:

https://github.com/GoogleCloudPlatform/kubernetes/blob/cd6daae014f8d3715f7305b49139df7210f1902a/pkg/registry/service/rest.go#L97-L121

markturansky committed:
Added pending phase for volumes.  added defaults for PV/PVC.  refactored to better phase transitioning in control loops

markturansky commented:
@thockin  feedback addressed and some refactorings introduced to aid readability:

* Refactor phase transitions in sync methods.  The phases are clearly laid out.  Many thanks to @ironcladlou for his advice here.
* Added default "pending" phase for volumes to aid phase changes in sync methods.
* Added watch to integration test to wait for expected changes to occur 
* Factored out the informat add/update funcs to provide good locking between my two controllers (volumes and claims both working on the shared volumeIndex)
* Also rebased

markturansky commented:
Rebased again and all green.

markturansky commented:
Rebased and green :)

markturansky commented:
"Rebased and green" is going to be my next album name.

chrismoos commented on pkg/api/validation/validation.go:
```diff
@@ -319,6 +319,11 @@ func validateSource(source *api.VolumeSource) errs.ValidationErrorList {
 	}
+	if source.PersistentVolumeClaimVolumeSource != nil {
+		numVolumes++
+		allErrs = append(allErrs, validatePersistentClaimVolumeSource(source.PersistentVolumeClaimVolumeSource).Prefix("glusterfs")...)
```
The prefix on this error is incorrect, should be persistentVolumeClaim.

markturansky replied:
good catch.  thanks.

markturansky committed:
fixed prefix name in validation

markturansky commented:
Rebased again.  

@thockin PTAL?

thockin commented:
On it

thockin commented:
I'm going to do a review of this as-is, but there's sort of a monster problem: This should use a "binding" sub-resource, as with pods.  We've drawn the analogy of claims and PVs being like pods and nodes - we should follow the same pattern.  That seems like a significant change, so I am willing to proceed on this path as long as we have a strong promise to fix it ASAP.  This is, strictly speaking, an API change (albeit constrained to ~1 consumer - you).

As per our IRC chat, there are a lot of things this might change:
* Do we bind the Claim to the PV or vice-versa?
* Do we move the Claim.VolumeRef to Spec?
* Do we allow users to ask for a specific PV?

thockin commented on pkg/api/testing/fuzzer.go:
```diff
@@ -201,6 +201,14 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
 		},
+		func(pv *api.PersistentVolume, c fuzz.Continue) {
+			c.FuzzNoCustom(pv) // fuzz self without calling this function again
+			pv.Status.Phase = api.VolumePending
```
These are not fuzzing.  You should pick a random number [0-num_phases) and index into a table of phase strings.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,359 @@
+		syncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    binder.addClaim,
+			UpdateFunc: binder.updateClaim,
```
no deleteClaim hook?  Can you comment why?

markturansky replied:
comments added but this line not touched (no collapse)

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,359 @@
+
+	if currentPhase != nextPhase {
+		volume.Status.Phase = nextPhase
+		volume, err := binderClient.UpdatePersistentVolumeStatus(volume)
```
Will this trigger an immediate new event which will run this function again?  Comment?

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,359 @@
+				// bound and active.  Build claim status as needed.
+				if claim.Status.VolumeRef == nil {
+					syncClaimStatus(binderClient, volume, claim)
+				}
```
don't you want to validate that the VolumeRef points back to this volume?  Or is that done somewhere else?  Comment

markturansky replied:
Do you mean that I should have code that validates that the volume just bound to is the volume passed into the func?  Or do you mean that the claim should check that its bound volume still exists?

If the latter, that would go in the control loop for the claim.  I had a check like that very early on but thought it would be an egregious violation of protocol to have a volume deleted from beneath a claim and that the error would come up through the pod.

thockin replied:
It was a question, not a directive.  This tests that "I" have a reference to "him", and that "he" exists, and that "he" has a reference to "someone".  I was asking if it is worthwhile to check that "someone" is "me" ?

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,359 @@
+	nextPhase := currentPhase
+
+	switch currentPhase {
+
```
guess what I;ll say about this blank line?

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,359 @@
+
+		// make a binding reference to the claim.
+		// triggers update of the volume in this controller, which builds claim status
+		volume.Spec.ClaimRef = claimRef
```
As per the binding sub-resource discussion, I sense that we have this backwards.  PersVolume.Status should host a ClaimRef and PersVolumeClaim.Spec should host a VolumeName (not ref, I think).

thockin commented:
Looks good overall.  Let's have the discussion about sub-resource.  I think that doing it right means flipping the current status/spec arrangement and that is a tedious and breaking API change.

As per IRC - maybe we can flag gate this entirely?  get this committed and off-the-books, let your folks proceed with testing, but force anyone who wants to use it to say --enable_alpha_persistent_volumes_support or something

markturansky committed:
addressed feedback.  added opt-in cmd line flag

markturansky commented:
Immediate feedback addressed and added flag to optionally start the binder in controller-manager.

Regarding the proposed change for binding, I will create a new PR as a follow-on task.

thockin commented on pkg/volumeclaimbinder/persistent_volume_claim_binder.go:
```diff
@@ -0,0 +1,365 @@
+		if volume.Spec.ClaimRef == nil {
+			return fmt.Errorf("PersistentVolume[%s] expected to be bound but found nil claimRef: %+v", volume)
+		} else {
+			// TODO: implement Recycle method on plugins
```
Just to be clear, at this point the admin must remove and recreate the PV object or set Phase manually, yes?  There's no other way to revive it?

markturansky replied:
Yes, the current reclamation process is manual.

thockin commented:
LGTM - do you want to fix the last nit or not worthwhile?

markturansky commented:
@thockin i'm good as-is and will start looking to the Binding subresource.
